File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/connection/builder/GivenANullLibrary/WhenScanningForUrls.java
package com.lasthopesoftware.bluewater.client.connection.builder.GivenANullLibrary;

import com.lasthopesoftware.bluewater.client.connection.builder.UrlScanner;
import com.lasthopesoftware.bluewater.client.connection.builder.lookup.LookupServers;
import com.lasthopesoftware.bluewater.client.connection.okhttp.ProvideOkHttpClients;
import com.lasthopesoftware.bluewater.client.connection.testing.TestConnections;
import com.lasthopesoftware.bluewater.shared.promises.extensions.FuturePromise;
import com.lasthopesoftware.resources.strings.EncodeToBase64;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

public class WhenScanningForUrls {

	private static IllegalArgumentException illegalArgumentException;

	@BeforeClass
	public static void before() throws InterruptedException, ExecutionException {
		final UrlScanner urlScanner = new UrlScanner(
			mock(EncodeToBase64.class),
			mock(TestConnections.class),
			mock(LookupServers.class),
			mock(ProvideOkHttpClients.class));

		try {
			new FuturePromise<>(urlScanner.promiseBuiltUrlProvider(null)).get();
		} catch (ExecutionException e) {
			if (e.getCause() instanceof IllegalArgumentException)
				illegalArgumentException = (IllegalArgumentException) e.getCause();
			else
				throw e;
		}
	}

	@Test
	public void thenAnIllegalArgumentExceptionIsThrown() {
		assertThat(illegalArgumentException).isNotNull();
	}

	@Test
	public void thenTheExceptionMentionsTheLibrary() {
		assertThat(illegalArgumentException.getMessage()).isEqualTo("The library cannot be null");
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/connection/builder/GivenANullLibrary/WhenScanningForUrls.kt
package com.lasthopesoftware.bluewater.client.connection.builder.GivenANullLibrary

import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.builder.MissingConnectionSettingsException
import com.lasthopesoftware.bluewater.client.connection.builder.UrlScanner
import com.lasthopesoftware.bluewater.client.connection.settings.LookupConnectionSettings
import com.lasthopesoftware.bluewater.shared.promises.extensions.toFuture
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.BeforeClass
import org.junit.Test
import java.util.concurrent.ExecutionException

class WhenScanningForUrls {

	companion object {
		private var exception: MissingConnectionSettingsException? = null

		@BeforeClass
		@JvmStatic
		fun before() {
			val connectionSettingsLookup = mockk<LookupConnectionSettings>()
			every { connectionSettingsLookup.lookupConnectionSettings(any()) } returns Promise.empty()

			val urlScanner = UrlScanner(
				mockk(),
				mockk(),
				mockk(),
				connectionSettingsLookup,
				mockk()
			)
			try {
				urlScanner.promiseBuiltUrlProvider(LibraryId(32)).toFuture().get()
			} catch (e: ExecutionException) {
				exception = e.cause as? MissingConnectionSettingsException ?: throw e
			}
		}
	}

	@Test
	fun thenTheCorrectExceptionIsThrown() {
		assertThat(exception).isNotNull
	}

	@Test
	fun thenTheExceptionMentionsTheLibrary() {
		assertThat(exception?.message).isEqualTo("Connection settings were not found for ${LibraryId(32)}")
	}
}
